panic: tview.(*Table).Draw.func1

I am getting a panic in tview. It looks like it might be in table.go line [469](https://github.com/rivo/tview/blob/6614b16d9037759065eac63d8fb3d8408e63e222/table.go#L469), but it is really hard to troubleshoot because of the way the panic is being printed.

![panic](https://user-images.githubusercontent.com/4334144/43463566-7eca8cbc-949e-11e8-90e7-6653949efb94.png)

(Click image for bigger view)

@rivo This issue ended up happening again. Just not quite as often. I had to edit table.go to capture the panic.

panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x6a0440]

goroutine 6 [running]:
github.com/rivo/tview.(*Table).Draw.func1(0x8232e0, 0xc042126d80)
        D:/gowork/src/github.com/rivo/tview/table.go:459 +0x86
panic(0x77b820, 0x9ccda0)
        C:/Go/src/runtime/panic.go:502 +0x237
github.com/rivo/tview.(*Table).Draw.func2(...)
        D:/gowork/src/github.com/rivo/tview/table.go:478
github.com/rivo/tview.(*Table).Draw(0xc042059360, 0x8232e0, 0xc042126d80)
        D:/gowork/src/github.com/rivo/tview/table.go:792 +0x1eb0
github.com/rivo/tview.(*Pages).Draw(0xc0420f9d40, 0x8232e0, 0xc042126d80)
        D:/gowork/src/github.com/rivo/tview/pages.go:246 +0xfd
github.com/rivo/tview.(*Grid).Draw(0xc042122090, 0x8232e0, 0xc042126d80)
        D:/gowork/src/github.com/rivo/tview/grid.go:631 +0x150b
github.com/rivo/tview.(*Pages).Draw(0xc0420f9d10, 0x8232e0, 0xc042126d80)
        D:/gowork/src/github.com/rivo/tview/pages.go:246 +0xfd
github.com/rivo/tview.(*Application).Draw(0xc042108af0, 0x0)
        D:/gowork/src/github.com/rivo/tview/application.go:259 +0xda
github.com/joegrasse/goburrow/goburrow/ui.NewApp.func2(0xc0421092d0)
        D:/gowork/src/github.com/joegrasse/goburrow/goburrow/ui/app.go:184 +0x4f
created by github.com/joegrasse/goburrow/goburrow/ui.NewApp
        D:/gowork/src/github.com/joegrasse/goburrow/goburrow/ui/app.go:181 +0xdc0

--- a/application.go
+++ b/application.go
@@ -46,11 +46,20 @@ type Application struct {
 
 	// Halts the event loop during suspended mode.
 	suspendMutex sync.Mutex
+
+	// Used to send screen events from separate goroutine to main event loop
+	events chan tcell.Event
+
+	// Used to send primitive updates from separate goroutines to the main event loop
+	updates chan func()
 }
 
 // NewApplication creates and returns a new application.
 func NewApplication() *Application {
-	return &Application{}
+	return &Application{
+		events:  make(chan tcell.Event, 100),
+		updates: make(chan func(), 100),
+	}
 }
 
 // SetInputCapture sets a function which captures all key events before they are
@@ -136,61 +145,78 @@ func (a *Application) Run() error {
 	a.Unlock()
 	a.Draw()
 
-	// Start event loop.
-	for {
-		// Do not poll events during suspend mode
-		a.suspendMutex.Lock()
-		a.RLock()
-		screen := a.screen
-		a.RUnlock()
-		if screen == nil {
-			a.suspendMutex.Unlock()
-			break
-		}
-
-		// Wait for next event.
-		event := a.screen.PollEvent()
-		a.suspendMutex.Unlock()
-		if event == nil {
-			// The screen was finalized. Exit the loop.
-			break
-		}
-
-		switch event := event.(type) {
-		case *tcell.EventKey:
-			a.RLock()
-			p := a.focus
-			a.RUnlock()
-
-			// Intercept keys.
-			if a.inputCapture != nil {
-				event = a.inputCapture(event)
-				if event == nil {
-					break // Don't forward event.
-				}
-			}
-
-			// Ctrl-C closes the application.
-			if event.Key() == tcell.KeyCtrlC {
-				a.Stop()
-			}
-
-			// Pass other key events to the currently focused primitive.
-			if p != nil {
-				if handler := p.InputHandler(); handler != nil {
-					handler(event, func(p Primitive) {
-						a.SetFocus(p)
-					})
-					a.Draw()
-				}
-			}
-		case *tcell.EventResize:
+	// Separate loop to wait for screen events
+	go func() {
+		for {
+			// Do not poll events during suspend mode
+			a.suspendMutex.Lock()
 			a.RLock()
 			screen := a.screen
 			a.RUnlock()
-			screen.Clear()
+			if screen == nil {
+				a.suspendMutex.Unlock()
+				// send signal to stop main event loop
+				a.QueueEvent(nil)
+				break
+			}
+
+			// Wait for next event.
+			a.QueueEvent(screen.PollEvent())
+			a.suspendMutex.Unlock()
+		}
+	}()
+
+	// Start event loop.
+loop:
+	for {
+		select {
+		case event := <-a.events:
+			if event == nil {
+				// The screen was finalized. Exit the loop.
+				break loop
+			}
+
+			switch event := event.(type) {
+			case *tcell.EventKey:
+				a.RLock()
+				p := a.focus
+				a.RUnlock()
+
+				// Intercept keys.
+				if a.inputCapture != nil {
+					event = a.inputCapture(event)
+					if event == nil {
+						break loop // Don't forward event.
+					}
+				}
+
+				// Ctrl-C closes the application.
+				if event.Key() == tcell.KeyCtrlC {
+					a.Stop()
+				}
+
+				// Pass other key events to the currently focused primitive.
+				if p != nil {
+					if handler := p.InputHandler(); handler != nil {
+						handler(event, func(p Primitive) {
+							a.SetFocus(p)
+						})
+						a.Draw()
+					}
+				}
+			case *tcell.EventResize:
+				a.RLock()
+				screen := a.screen
+				a.RUnlock()
+				screen.Clear()
+				a.Draw()
+			}
+
+		case updater := <-a.updates:
+			updater()
 			a.Draw()
 		}
+
 	}
 
 	return nil
@@ -403,4 +429,16 @@ func (a *Application) GetFocus() Primitive {
 	a.RLock()
 	defer a.RUnlock()
 	return a.focus
+}
+
+// QueueUpdate is used to synchronize changes to primitives by carrying an update function from separate goroutine to the Application event loop via channel
+func (a *Application) QueueUpdate(f func()) *Application {
+	a.updates <- f
+	return a
+}
+
+// QueueEvent takes an Event instance and sends it to the Application event loop via channel
+func (a *Application) QueueEvent(e tcell.Event) *Application {
+	a.events <- e
+	return a
 }

